_posts/2020-02-17-constructor-injection.md

```diff
@@ -0,0 +1,189 @@
+---
+title: Why You should use Constructor Injection in Spring
+categories: [spring-boot]
+date: 2020-02-16 05:00:00 +1100
+modified: 2020-02-16 05:00:00 +1100
+author: default
+excerpt: 'Dependency injection is a one of the common approaches to implement loose coupling among the classes in an application. It ensures that classes are independent and the required dependencies are provided by an external framework.'
+image:
+auto: 0058-motorway-junction
+---
+Dependency injection is a one of the common approaches to implement loose coupling among the classes in an application. It ensures that classes are independent and the required dependencies are provided by an external framework.
+### What is dependency injection?
+
+`Dependency` - Every object of a class requires some dependencies to perform its operation. These dependencies can be objects of other classes.
+
+`Injection` - Providing required dependencies to object.
+
+Thus dependency injection helps in implementing inversion of control meaning, the responsibility of object creation and injecting the dependencies is given to the framework (i.e. Container in Spring) instead of the class creating the dependent objects by itself.
+
+We can implement this in three ways:
+
+ 1. Constructor based injection
+ 2. Field based injection
+ 3. Setter based injection
+
+You can find the code of the example demonstrated in the below git URL.
+{% include github-project.html url="https://github.com/vasudhavenkatesan/DependencyInjectionExample" %}
+
+# Constructor Injection
+
+In constructor based injection, the dependencies required for the class are provided in the constructor, i.e. the required dependencies are passed as arguments to the constructor.
+
+```java
+@Component
+public class Cake {
+
+	private static Logger LOGGER = LoggerFactory.getLogger(Cake.class);
+
+
+	private Flavor flavor;
+
+	public Cake(Flavor flavor) throws IllegalAccessException {
+		
+		//check if the required dependency is not null
+		if (flavor != null) {
+			this.flavor = flavor;
+			LOGGER.info("Flavor from Constructor Injection : " + flavor);
+		} else {
+			throw new IllegalArgumentException("Cake cannot be created with null flavor object");
+		}
+	}
+	...
+}
+```
+
+After Spring 4.3, **it is optional to specify `@Autowired` if the class has only one constructor defined**. However, if we have a class with multiple constructors, we need to explicitly mention `@Autowired` annotation to any one of the constructors for the container to create beans.
+
+# Setter/Field based injection
+
+In setter-based injection, we can provide the required dependencies as field parameters to the class and values are set using the setter methods of the properties. We need to annotate the setter methods with `@Autowired`  annotation. Similarly, in field-based injection, Spring assigns the required dependencies directly to the fields on annotating with `@Autowired` annotation. 
```

Comment:
I have mentioned it on the text that we can use `@Autowired` on getters or setters. In the example I have shown it using setters because I dont get the point of using it in getters. Also included example having `@Autowired` for same field using setter and field injection.